main/test/ca/uwaterloo/flix/language/phase/TestSafety.scala

```diff
@@ -180,4 +181,83 @@ class TestSafety extends FunSuite with TestUtils {
     expectError[IllegalRelationalUseOfLatticeVariable](result)
   }
 
+  test("TestIllegalDerivation.01") {
+    val input =
+      """
+        |def f(): ##java.lang.Thread & Impure = object ##java.lang.Thread {}
+      """.stripMargin
+    val result = compile(input, Options.TestWithLibNix)
+    expectError[SafetyError.IllegalObjectDerivation](result)
+  }
+
+  test("TestInvalidThis.01") {
+    val input =
+      """
+        |def f(): ##java.lang.Runnable & Impure = 
+        |  object ##java.lang.Runnable {
+        |    def run(): Unit & Impure = ()
+        |  }
+      """.stripMargin
+    val result = compile(input, Options.TestWithLibNix)
+    expectError[SafetyError.IllegalThisType](result)
+  }
+
+  test("TestInvalidThis.02") {
+    val input =
+      """
+        |def f(): ##java.lang.Runnable & Impure = 
+        |  object ##java.lang.Runnable {
+        |    def run(_this: Int32): Unit & Impure = ()
+        |  }
+      """.stripMargin
+    val result = compile(input, Options.TestWithLibNix)
+    expectError[SafetyError.IllegalThisType](result)
+  }
+
+  test("TestInvalidThis.03") {
```

Comment:
We keep all our positive tests (i.e. the ones that do not fail) in flix files, e.g. in `Test.Exp.Jvm.NewObject.flix`. The reason is performance, by having multiple test files in the same Flix file we can execute everything a lot quicker.